tools/clang/lib/AST/HlslTypes.cpp

```diff
@@ -645,15 +606,62 @@ bool IsHLSLResourceType(clang::QualType type) {
   return false;
 }
 
-bool IsHLSLNodeInputType(clang::QualType type) {
+static HLSLNodeRecordAttr *getNodeAttr(clang::QualType type) {
   if (const RecordType *RT = type->getAs<RecordType>()) {
-    StringRef name = RT->getDecl()->getName();
-    if (name == "EmptyNodeInput" || name == "DispatchNodeInputRecord" ||
-        name == "RWDispatchNodeInputRecord" ||
-        name == "GroupNodeInputRecords" || name == "RWGroupNodeInputRecords" ||
-        name == "ThreadNodeInputRecord" || name == "RWThreadNodeInputRecord")
-      return true;
+    if (const auto *Spec =
+            dyn_cast<ClassTemplateSpecializationDecl>(RT->getDecl()))
+      if (const auto *Template =
+              dyn_cast<ClassTemplateDecl>(Spec->getSpecializedTemplate()))
+        return Template->getTemplatedDecl()->getAttr<HLSLNodeRecordAttr>();
+    if (const auto *Decl = dyn_cast<CXXRecordDecl>(RT->getDecl()))
+      return Decl->getAttr<HLSLNodeRecordAttr>();
+  }
+  return nullptr;
+}
+
+DXIL::NodeIOKind GetNodeIOType(clang::QualType type) {
+  if (const HLSLNodeRecordAttr *Attr = getNodeAttr(type)) {
+    switch (Attr->getType()) {
+    case HLSLNodeRecordAttr::DispatchNodeInputRecord:
+      return DXIL::NodeIOKind::DispatchNodeInputRecord;
+    case HLSLNodeRecordAttr::RWDispatchNodeInputRecord:
+      return DXIL::NodeIOKind::RWDispatchNodeInputRecord;
+    case HLSLNodeRecordAttr::GroupNodeInputRecords:
+      return DXIL::NodeIOKind::GroupNodeInputRecords;
+    case HLSLNodeRecordAttr::RWGroupNodeInputRecords:
+      return DXIL::NodeIOKind::RWGroupNodeInputRecords;
+    case HLSLNodeRecordAttr::ThreadNodeInputRecord:
+      return DXIL::NodeIOKind::ThreadNodeInputRecord;
+    case HLSLNodeRecordAttr::RWThreadNodeInputRecord:
+      return DXIL::NodeIOKind::RWThreadNodeInputRecord;
+    case HLSLNodeRecordAttr::EmptyNodeInput:
+      return DXIL::NodeIOKind::EmptyInput;
+    case HLSLNodeRecordAttr::NodeOutput:
+      return DXIL::NodeIOKind::NodeOutput;
+    case HLSLNodeRecordAttr::EmptyNodeOutput:
+      return DXIL::NodeIOKind::EmptyOutput;
+    case HLSLNodeRecordAttr::NodeOutputArray:
+      return DXIL::NodeIOKind::NodeOutputArray;
+    case HLSLNodeRecordAttr::EmptyNodeOutputArray:
+      return DXIL::NodeIOKind::EmptyOutputArray;
+    case HLSLNodeRecordAttr::GroupNodeOutputRecords:
+      return DXIL::NodeIOKind::GroupNodeOutputRecords;
+    case HLSLNodeRecordAttr::ThreadNodeOutputRecords:
+      return DXIL::NodeIOKind::ThreadNodeOutputRecords;
+    }
   }
+  return DXIL::NodeIOKind::Invalid;
+}
+
+bool IsHLSLNodeInputType(clang::QualType type) {
+  if (const HLSLNodeRecordAttr *Attr = getNodeAttr(type))
+    return Attr->getType() == HLSLNodeRecordAttr::EmptyNodeInput ||
```

Comment:
It's a shame this isn't using the native `NodeIOKind` values, since those are organized such that questions like this can be answered more easily (`kind & NodeIOFlags::Input`).

Reply:
I could rewrite it that way if you prefer, but I liked the explicitness of the specific types.